src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1143 @@
+        ldpp_dout(dpp, 10) << __func__ << "::++ACK::ctl=" << ctl << "::ret=" << ret << dendl;
+      } catch (buffer::error& err) {
+        ldpp_dout(dpp, 1) << __func__ << "::failed decoding notify acks" << dendl;
+        return -EINVAL;
+      }
+      if (ret != 0) {
+        ldpp_dout(dpp, 1) << __func__ << "::Bad notify ack, ret=" << ret
+                          << "::err=" << cpp_strerror(ret) << dendl;
+        return ret;
+      }
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::" << get_urgent_msg_names(urgent_msg)
+                       << " finished successfully!" << dendl;
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  // command-line called from radosgw-admin.cc
+  int cluster::dedup_restart_scan(rgw::sal::RadosStore *store,
+                                  dedup_req_type_t dedup_type,
+                                  const DoutPrefixProvider *dpp)
+  {
+    librados::IoCtx ioctx;
+    int ret = init_dedup_pool_ioctx(store->getRados(), dpp, ioctx);
+    if (ret != 0) {
+      return ret;
+    }
+
+    dedup_epoch_t old_epoch;
+    // store the previous epoch for cmp-swap
+    ret = get_epoch(ioctx, dpp, &old_epoch, __func__);
+    if (ret != 0) {
+      return ret;
+    }
+
+    // first abort all dedup work!
+    ret = dedup_control(store, dpp, URGENT_MSG_ABORT);
+    if (ret != 0) {
+      return ret;
+    }
+
+    ldpp_dout(dpp, 10) << __func__ << dedup_type << dendl;
+#ifdef FULL_DEDUP_SUPPORT
+    ceph_assert(dedup_type == dedup_req_type_t::DEDUP_TYPE_DRY_RUN ||
+                dedup_type == dedup_req_type_t::DEDUP_TYPE_FULL);
+#else
+    ceph_assert(dedup_type == dedup_req_type_t::DEDUP_TYPE_DRY_RUN);
+#endif
+    ret = swap_epoch(dpp, ioctx, &old_epoch, dedup_type, 0, 0);
+    if (ret == 0) {
+      ldpp_dout(dpp, 10) << __func__ << "::Epoch object was reset" << dendl;
+      return dedup_control(store, dpp, URGENT_MSG_RESTART);
+    }
+    else {
+      return ret;
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  bool cluster::can_start_new_scan(rgw::sal::RadosStore *store)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::epoch=" << d_epoch_time << dendl;
+    librados::IoCtx ioctx;
+    int ret = init_dedup_pool_ioctx(store->getRados(), dpp, ioctx);
+    if (ret != 0) {
+      return ret;
+    }
+
+    dedup_epoch_t new_epoch;
+    if (get_epoch(ioctx, dpp, &new_epoch, nullptr) != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::No Epoch Object::"
+                        << "::scan can be restarted!\n\n\n" << dendl;
+      // no epoch object exists -> we should start a new scan
+      return true;
+    }
+
+    if (new_epoch.time <= d_epoch_time) {
+      if (new_epoch.time == d_epoch_time) {
+        ldpp_dout(dpp, 10) << __func__ << "::Epoch hasn't change - > Do not restart scan!!" << dendl;
+      }
+      else {
+        ldpp_dout(dpp, 1) << __func__ << " ::Do not restart scan!\n    epoch="
```

Comment:
Suggest no ! in these messages, multiple above and below throughout files.

Reply:
why? it is meant to catch the attention of the one looking in the log